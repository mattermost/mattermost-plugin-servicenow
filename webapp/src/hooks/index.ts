import {Store, Action} from 'redux';

import {GlobalState} from 'mattermost-redux/types/store';

export default class Hooks {
    store: Store<GlobalState, Action<Record<string, unknown>>>

    constructor(store: Store<GlobalState, Action<Record<string, unknown>>>) {
        this.store = store;
    }
}
