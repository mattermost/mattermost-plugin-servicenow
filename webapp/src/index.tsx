import React from 'react';
import {Store, Action} from 'redux';

import {GlobalState} from 'mattermost-redux/types/store';

// eslint-disable-next-line import/no-unresolved
import {PluginRegistry} from 'types/mattermost-webapp';

import {ServiceNowIcon} from '@brightscout/mattermost-ui-library';

import reducer from 'reducers';

import Rhs from 'containers/Rhs';
import AddSubscription from 'containers/addOrEditSubscriptions/addSubscription';
import EditSubscription from 'containers/addOrEditSubscriptions/editSubscription';

import Constants from 'plugin_constants';

import DownloadButton from 'components/admin_settings/download_button';
import {handleConnect, handleDisconnect, handleOpenAddSubscriptionModal, handleOpenEditSubscriptionModal, handleSubscriptionDeleted} from 'websocket';

import App from './app';

import manifest from './manifest';

import './styles/main.scss';

export default class Plugin {
    // eslint-disable-next-line @typescript-eslint/no-unused-vars, @typescript-eslint/no-empty-function
    public async initialize(registry: PluginRegistry, store: Store<GlobalState, Action<Record<string, unknown>>>) {
        // @see https://developers.mattermost.com/extend/plugins/webapp/reference/
        registry.registerReducer(reducer);
        registry.registerRootComponent(AddSubscription);
        registry.registerRootComponent(EditSubscription);
        registry.registerRootComponent(App);
        const {id, toggleRHSPlugin} = registry.registerRightHandSidebarComponent(Rhs, Constants.RightSidebarHeader);
        registry.registerChannelHeaderButtonAction(<ServiceNowIcon/>, () => store.dispatch(toggleRHSPlugin), null, Constants.ChannelHeaderTooltipText);
        registry.registerAdminConsoleCustomSetting('ServiceNowUpdateSetDownload', DownloadButton);

        registry.registerWebSocketEventHandler(`custom_${manifest.id}_connect`, handleConnect(store, id));
        registry.registerWebSocketEventHandler(`custom_${manifest.id}_disconnect`, handleDisconnect(store));
        registry.registerWebSocketEventHandler(`custom_${manifest.id}_add_subscription`, handleOpenAddSubscriptionModal(store));
        registry.registerWebSocketEventHandler(`custom_${manifest.id}_edit_subscription`, handleOpenEditSubscriptionModal(store));
        registry.registerWebSocketEventHandler(`custom_${manifest.id}_subscription_deleted`, handleSubscriptionDeleted(store, id));
    }
}

declare global {
    interface Window {
        registerPlugin(id: string, plugin: Plugin): void
    }
}

window.registerPlugin(manifest.id, new Plugin());
