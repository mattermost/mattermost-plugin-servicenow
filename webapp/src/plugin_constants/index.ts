/**
* Keep all plugin related constants here
*/
export enum ToggleSwitchLabelPositioning {
    Left = 'left',
    Right = 'right',
}

export const UPLOAD_SET_FILENAME = 'servicenow_for_mattermost_notifications_v2.3.xml';
export const CONNECT_ACCOUNT_LINK = '/oauth2/connect';
export const SERVICENOW_ICON_URL = 'servicenow-icon.png';
export const REQUEST_BASE_URL = 'com.glideapp.servicecatalog_cat_item_view.do?v=1&sysparm_id=';
export const SERVICENOW_SUBSCRIPTIONS_TABLE = 'x_830655_mm_std_servicenow_for_mattermost_subscriptions';

const MMCSRF = 'MMCSRF';
const HeaderCSRFToken = 'X-CSRF-Token';
const MMUSERID = 'MMUSERID';
const RightSidebarHeader = 'ServiceNow';
const RhsSubscritpions = 'Subscriptions';
const ShareRecordButton = 'Share';
const RhsToggleLabel = 'Show all subscriptions';
const ChannelPanelToggleLabel = 'Subscribe to the new incident';
const InvalidAutoCompleteValueMsg = 'Invalid value, please select a value from the suggestions.';
const ChannelHeaderTooltipText = 'ServiceNow';
const DefaultCharThresholdToShowSuggestions = 3;
const DefaultPage = 0;
const DefaultPageSize = 20;
const ApiErrorIdNotConnected = 'not_connected';
const ApiErrorIdRefreshTokenExpired = 'refresh_token_expired';
const ApiErrorIdSubscriptionsNotConfigured = 'subscriptions_not_configured';
const ApiErrorIdSubscriptionsUnauthorized = 'subscriptions_not_authorized';
const GeneralErrorMessage = 'Something went wrong';
const GeneralErrorSubtitleForUser = 'Please contact your system administrator.';
const GeneralErrorSubtitleForAdmin = 'Please check the server logs.';
const SubscriptionAddedMsg = 'Subscription added successfully!';
const SubscriptionUpdatedMsg = 'Subscription updated successfully!';
const DeleteSubscriptionHeading = 'Confirm Subscription Delete';
const DeleteSubscriptionMsg = 'Are you sure you want to delete the subscription?';
const RecordSharedMsg = 'Record shared successfully!';
const StateUpdatedMsg = 'State updated successfully!';
const IncidentCreatedMsg = 'Incident created successfully!';
const CharThresholdToSuggestChannel = 0;
const CharThresholdToSuggestRequest = 4;
const RequiredMsg = 'Required';
const NoSubscriptionPresent = 'No more subscriptions present.';
const CommentsHeading = 'Comments';
const NoCommentsPresent = 'No more comments present.';
const CommentsNotFound = 'No comments found.';
const EmptyFieldsInServiceNow = 'N/A';
const ServiceNowSysIdRegex = '[0-9a-f]{32}';
const RequestButtonText = 'Submit Request on ServiceNow';
const RequestButtonRedirectText = 'You will be redirected to ServiceNow to complete this request';
const DefaultPerPageParam = 10;
const DebounceFunctionTimeLimit = 500;
const MaxShortDescriptionCharactersView = 75;
const MaxShortDescriptionLimit = 160;

export enum ModalIds {
    ADD_SUBSCRIPTION = 'addSubscription',
    EDIT_SUBSCRIPTION = 'editSubscription',
    SHARE_RECORD = 'shareRecord',
    ADD_OR_VIEW_COMMENTS = 'addOrViewComments',
    UPDATE_STATE = 'updateState',
    CREATE_INCIDENT = 'createIncident',
    CREATE_REQUEST = 'createRequest',
}

export enum SubscriptionEvents {
    CREATED = 'created',
    STATE = 'state',
    PRIORITY = 'priority',
    COMMENTED = 'commented',
    ASSIGNED_TO = 'assigned_to',
    ASSIGNMENT_GROUP = 'assignment_group',
}

export enum SubscriptionType {
    RECORD = 'record',
    BULK = 'object',
}

export enum RecordType {
    INCIDENT = 'incident',
    PROBLEM = 'problem',
    CHANGE_REQUEST = 'change_request',
    KNOWLEDGE = 'kb_knowledge',
    TASK = 'task',
    CHANGE_TASK = 'change_task',
    FOLLOW_ON_TASK = 'cert_follow_on_task',
}

export enum SupportedFilters {
    ASSIGNMENT_GROUP = 'assignment_group',
    SERVICE = 'business_service',
}

export const SupportedFiltersMap: Record<SupportedFilters, string> = {
    [SupportedFilters.ASSIGNMENT_GROUP]: '"assignment_group"',
    [SupportedFilters.SERVICE]: '"business_service"',
};

export const SubscriptionEventsMap: Record<string, SubscriptionEvents> = {
    created: SubscriptionEvents.CREATED,
    state: SubscriptionEvents.STATE,
    priority: SubscriptionEvents.PRIORITY,
    commented: SubscriptionEvents.COMMENTED,
    assigned_to: SubscriptionEvents.ASSIGNED_TO,
    assignment_group: SubscriptionEvents.ASSIGNMENT_GROUP,
};

const SubscriptionsConfigErrorTitle = 'It seems that subscriptions for ServiceNow have not been configured properly.';
const SubscriptionsConfigErrorSubtitleForUser = 'Please contact your system administrator to configure the subscriptions by following the instructions given by the plugin.';
const SubscriptionsConfigErrorSubtitleForAdmin = 'To enable subscriptions, you have to download the update set provided by the plugin and upload that in ServiceNow. The update set is available in the plugin configuration settings or you can download it by clicking the button below. The instructions for uploading the update set are available in the plugin\'s documentation and also can be viewed by running the "/servicenow help" command.';
const SubscriptionsUnauthorizedErrorTitle = 'It seems that you are not authorized to manage subscriptions in ServiceNow.';
const SubscriptionsUnauthorizedErrorSubtitleForUser = 'Please contact your system administrator to authorize you for managing subscriptions.';
const SubscriptionsUnauthorizedErrorSubtitleForAdmin = 'Please follow the instructions for setting up user permissions available in the plugin\'s documentation. The instructions can also be viewed by running the "/servicenow help" command.';

// Used to get the `SubscriptionType` labels to show in the UI
export const SubscriptionTypeLabelMap: Record<SubscriptionType, string> = {
    [SubscriptionType.RECORD]: 'Record subscription',
    [SubscriptionType.BULK]: 'Bulk subscription',
};

// Used to get the `RecordType` labels to show in the UI
export const RecordTypeLabelMap: Record<RecordType, string> = {
    [RecordType.INCIDENT]: 'Incident',
    [RecordType.PROBLEM]: 'Problem',
    [RecordType.CHANGE_REQUEST]: 'Change Request',
    [RecordType.KNOWLEDGE]: 'Knowledge',
    [RecordType.TASK]: 'Task',
    [RecordType.CHANGE_TASK]: 'Change Task',
    [RecordType.FOLLOW_ON_TASK]: 'Follow On Task',
};

const recordTypeOptions: DropdownOptionType[] = [
    {
        label: RecordTypeLabelMap[RecordType.INCIDENT],
        value: RecordType.INCIDENT,
    },
    {
        label: RecordTypeLabelMap[RecordType.PROBLEM],
        value: RecordType.PROBLEM,
    },
    {
        label: RecordTypeLabelMap[RecordType.CHANGE_REQUEST],
        value: RecordType.CHANGE_REQUEST,
    },
];

const shareRecordTypeOptions: DropdownOptionType[] = recordTypeOptions.concat([
    {
        label: RecordTypeLabelMap[RecordType.KNOWLEDGE],
        value: RecordType.KNOWLEDGE,
    },
    {
        label: RecordTypeLabelMap[RecordType.TASK],
        value: RecordType.TASK,
    },
    {
        label: RecordTypeLabelMap[RecordType.CHANGE_TASK],
        value: RecordType.CHANGE_TASK,
    },
    {
        label: RecordTypeLabelMap[RecordType.FOLLOW_ON_TASK],
        value: RecordType.FOLLOW_ON_TASK,
    },
]);

export enum RecordDataConfigKeys {
    SHORT_DESCRIPTION = 'short_description',
    STATE = 'state',
    PRIORITY = 'priority',
    ASSIGNED_TO = 'assigned_to',
    ASSIGNMENT_GROUP = 'assignment_group',
    SERVICE = 'business_service',
}

export enum RecordDataConfigLabels {
    SHORT_DESCRIPTION = 'Short Description',
    STATE = 'State',
    PRIORITY = 'Priority',
    ASSIGNED_TO = 'Assigned To',
    ASSIGNMENT_GROUP = 'Assignment Group',
    SERVICE = 'Service',
}

export const SupportedFiltersLabelsMap: Record<SupportedFilters, string> = {
    [SupportedFilters.ASSIGNMENT_GROUP]: RecordDataConfigLabels.ASSIGNMENT_GROUP,
    [SupportedFilters.SERVICE]: RecordDataConfigLabels.SERVICE,
};

// Used in search records panel for rendering the key-value pairs of the record for showing the record details
const RecordDataLabelConfig: RecordDataLabelConfigType[] = [
    {
        key: RecordDataConfigKeys.SHORT_DESCRIPTION,
        label: RecordDataConfigLabels.SHORT_DESCRIPTION,
    }, {
        key: RecordDataConfigKeys.STATE,
        label: RecordDataConfigLabels.STATE,
    }, {
        key: RecordDataConfigKeys.PRIORITY,
        label: RecordDataConfigLabels.PRIORITY,
    }, {
        key: RecordDataConfigKeys.ASSIGNED_TO,
        label: RecordDataConfigLabels.ASSIGNED_TO,
    }, {
        key: RecordDataConfigKeys.ASSIGNMENT_GROUP,
        label: RecordDataConfigLabels.ASSIGNMENT_GROUP,
    }, {
        key: RecordDataConfigKeys.SERVICE,
        label: RecordDataConfigLabels.SERVICE,
    },
];

export enum SubscriptionFilters {
    ME = 'me',
    ANYONE = 'anyone',
}

export const DefaultSubscriptionFilters = {
    createdBy: SubscriptionFilters.ANYONE,
};

export const SubscriptionFilterCreatedByOptions = [
    {
        value: SubscriptionFilters.ME,
        label: 'Me',
    },
    {
        value: SubscriptionFilters.ANYONE,
        label: 'Anyone',
    },
];

export enum KnowledgeRecordDataConfigKeys {
    SHORT_DESCRIPTION = 'short_description',
    WORKFLOW_STATE = 'workflow_state',
    AUTHOR = 'author',
    CATEGORY = 'kb_category',
    KNOWLEDGE_BASE = 'kb_knowledge_base',
}

export enum KnowledgeRecordDataConfigLabels {
    SHORT_DESCRIPTION = 'Short Description',
    WORKFLOW_STATE = 'Workflow',
    AUTHOR = 'Author',
    CATEGORY = 'Category',
    KNOWLEDGE_BASE = 'Knowledge Base',
}

const KnowledgeRecordDataLabelConfig: RecordDataLabelConfigType[] = [
    {
        key: KnowledgeRecordDataConfigKeys.SHORT_DESCRIPTION,
        label: KnowledgeRecordDataConfigLabels.SHORT_DESCRIPTION,
    }, {
        key: KnowledgeRecordDataConfigKeys.WORKFLOW_STATE,
        label: KnowledgeRecordDataConfigLabels.WORKFLOW_STATE,
    }, {
        key: KnowledgeRecordDataConfigKeys.AUTHOR,
        label: KnowledgeRecordDataConfigLabels.AUTHOR,
    }, {
        key: KnowledgeRecordDataConfigKeys.CATEGORY,
        label: KnowledgeRecordDataConfigLabels.CATEGORY,
    }, {
        key: KnowledgeRecordDataConfigKeys.KNOWLEDGE_BASE,
        label: KnowledgeRecordDataConfigLabels.KNOWLEDGE_BASE,
    },
];

export enum RequestDataLabelConfigKey {
    SHORT_DESCRIPTION = 'short_description',
    PRICE = 'price',
    CATEGORY = 'title',
}

const RequestDataLabelConfig = [
    {
        key: RequestDataLabelConfigKey.SHORT_DESCRIPTION,
        label: 'Short Description',
    }, {
        key: RequestDataLabelConfigKey.CATEGORY,
        label: 'Category',
    }, {
        key: RequestDataLabelConfigKey.PRICE,
        label: 'Price',
    },
];

// Map subscription events to texts to be shown in the UI(on cards)
export const SubscriptionEventLabels: Record<SubscriptionEvents, string> = {
    [SubscriptionEvents.CREATED]: 'New record created',
    [SubscriptionEvents.STATE]: 'State changed',
    [SubscriptionEvents.PRIORITY]: 'Priority changed',
    [SubscriptionEvents.COMMENTED]: 'New comment',
    [SubscriptionEvents.ASSIGNED_TO]: 'Assigned to changed',
    [SubscriptionEvents.ASSIGNMENT_GROUP]: 'Assignment group changed',
};

export const DefaultIncidentImpactAndUrgencyOptions: DropdownOptionType[] = [
    {
        value: '1',
        label: '1 - High',
    },
    {
        value: '2',
        label: '2 - Medium',
    },
    {
        value: '3',
        label: '3 - Low',
    },
];

export const RecordTypesSupportingComments = new Set([
    RecordType.INCIDENT,
    RecordType.PROBLEM,
    RecordType.CHANGE_REQUEST,
    RecordType.TASK,
    RecordType.CHANGE_TASK,
    RecordType.FOLLOW_ON_TASK,
]);

export const RecordTypesSupportingStateUpdation = new Set([
    RecordType.INCIDENT,
    RecordType.TASK,
    RecordType.CHANGE_TASK,
    RecordType.FOLLOW_ON_TASK,
]);

export const KeysContainingLink = new Set([
    KnowledgeRecordDataConfigKeys.KNOWLEDGE_BASE,
    KnowledgeRecordDataConfigKeys.AUTHOR,
    KnowledgeRecordDataConfigKeys.CATEGORY,
    RecordDataConfigKeys.ASSIGNED_TO,
    RecordDataConfigKeys.ASSIGNMENT_GROUP,
    RecordDataConfigKeys.SERVICE,
    KnowledgeRecordDataConfigLabels.KNOWLEDGE_BASE,
    KnowledgeRecordDataConfigLabels.AUTHOR,
    KnowledgeRecordDataConfigLabels.CATEGORY,
    RecordDataConfigLabels.ASSIGNED_TO,
    RecordDataConfigLabels.ASSIGNMENT_GROUP,
    RecordDataConfigLabels.SERVICE,
]);

export type TypesContainingLink = KnowledgeRecordDataConfigKeys | RecordDataConfigKeys | KnowledgeRecordDataConfigLabels | RecordDataConfigLabels;

// Plugin api service (RTK query) configs
const pluginApiServiceConfigs: Record<ApiServiceName, PluginApiService> = {
    getChannels: {
        path: '/channels',
        method: 'GET',
        apiServiceName: 'getChannels',
    },
    searchRecords: {
        path: '/records',
        method: 'GET',
        apiServiceName: 'searchRecords',
    },
    getRecord: {
        path: '/records',
        method: 'GET',
        apiServiceName: 'getRecord',
    },
    createSubscription: {
        path: '/subscriptions',
        method: 'POST',
        apiServiceName: 'createSubscription',
    },
    fetchSubscriptions: {
        path: '/subscriptions',
        method: 'GET',
        apiServiceName: 'fetchSubscriptions',
    },
    editSubscription: {
        path: '/subscriptions',
        method: 'PATCH',
        apiServiceName: 'editSubscription',
    },
    deleteSubscription: {
        path: '/subscriptions',
        method: 'DELETE',
        apiServiceName: 'deleteSubscription',
    },
    getConfig: {
        path: '/config',
        method: 'GET',
        apiServiceName: 'getConfig',
    },
    getComments: {
        path: '/comments',
        method: 'GET',
        apiServiceName: 'getComments',
    },
    addComments: {
        path: '/comments',
        method: 'POST',
        apiServiceName: 'addComments',
    },
    shareRecord: {
        path: '/share',
        method: 'POST',
        apiServiceName: 'shareRecord',
    },
    getStates: {
        path: '/states',
        method: 'GET',
        apiServiceName: 'getStates',
    },
    updateState: {
        path: '/states',
        method: 'PATCH',
        apiServiceName: 'updateState',
    },
    searchItems: {
        path: '/catalog',
        method: 'GET',
        apiServiceName: 'searchItems',
    },
    getUsers: {
        path: '/users',
        method: 'GET',
        apiServiceName: 'getUsers',
    },
    createIncident: {
        path: '/incident',
        method: 'POST',
        apiServiceName: 'createIncident',
    },
    getFilterData: {
        path: '/filter',
        method: 'GET',
        apiServiceName: 'getFilterData',
    },
    getTableFields: {
        path: '/fields',
        method: 'GET',
        apiServiceName: 'getTableFields',
    },
    getConnectedUser: {
        path: '/connected',
        method: 'GET',
        apiServiceName: 'getConnectedUser',
    },
};

export const PanelDefaultHeights = {
    channelPanel: 210,
    subscriptionTypePanel: 195,
    recordTypePanel: 210,
    searchRecordPanel: 210,
    filtersPanel: 375,
    searchRecordPanelExpanded: 335,
    eventsPanel: 500,
    successPanel: 220,
    errorPanel: 267,
    panelHeader: 65,
};

export default {
    RightSidebarHeader,
    RhsSubscritpions,
    ShareRecordButton,
    UPLOAD_SET_FILENAME,
    SERVICENOW_ICON_URL,
    REQUEST_BASE_URL,
    SERVICENOW_SUBSCRIPTIONS_TABLE,
    pluginApiServiceConfigs,
    MMCSRF,
    HeaderCSRFToken,
    InvalidAutoCompleteValueMsg,
    RecordDataLabelConfig,
    KnowledgeRecordDataLabelConfig,
    RequestDataLabelConfig,
    MMUSERID,
    SubscriptionsConfigErrorTitle,
    SubscriptionsConfigErrorSubtitleForAdmin,
    SubscriptionsConfigErrorSubtitleForUser,
    ChannelHeaderTooltipText,
    RhsToggleLabel,
    ChannelPanelToggleLabel,
    DefaultCharThresholdToShowSuggestions,
    DefaultPage,
    DefaultPageSize,
    ApiErrorIdNotConnected,
    ApiErrorIdRefreshTokenExpired,
    ApiErrorIdSubscriptionsNotConfigured,
    ApiErrorIdSubscriptionsUnauthorized,
    SubscriptionsUnauthorizedErrorTitle,
    SubscriptionsUnauthorizedErrorSubtitleForUser,
    SubscriptionsUnauthorizedErrorSubtitleForAdmin,
    GeneralErrorMessage,
    GeneralErrorSubtitleForUser,
    GeneralErrorSubtitleForAdmin,
    SubscriptionAddedMsg,
    SubscriptionUpdatedMsg,
    DeleteSubscriptionHeading,
    DeleteSubscriptionMsg,
    RecordSharedMsg,
    StateUpdatedMsg,
    IncidentCreatedMsg,
    CharThresholdToSuggestChannel,
    CharThresholdToSuggestRequest,
    RequiredMsg,
    recordTypeOptions,
    shareRecordTypeOptions,
    NoSubscriptionPresent,
    CommentsHeading,
    NoCommentsPresent,
    CommentsNotFound,
    SubscriptionFilters,
    DefaultSubscriptionFilters,
    SubscriptionFilterCreatedByOptions,
    EmptyFieldsInServiceNow,
    ServiceNowSysIdRegex,
    RequestButtonText,
    RequestButtonRedirectText,
    DefaultPerPageParam,
    DebounceFunctionTimeLimit,
    MaxShortDescriptionCharactersView,
    MaxShortDescriptionLimit,
};
