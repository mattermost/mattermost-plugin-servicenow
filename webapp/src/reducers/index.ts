import {combineReducers} from 'redux';

import services from 'src/services';

import apiRequestCompletionReducer from './apiRequest';
import globalModalReducer from './globalModal';
import refetchReducer from './refetchState';
import connectedReducer from './connectedState';

export default combineReducers({
    apiRequestCompletionReducer,
    globalModalReducer,
    refetchReducer,
    connectedReducer,
    [services.reducerPath]: services.reducer,
});
