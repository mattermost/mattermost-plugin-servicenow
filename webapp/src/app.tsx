import React, {useEffect} from 'react';
import {useDispatch} from 'react-redux';

import usePluginApi from 'src/hooks/usePluginApi';

import Constants from 'src/plugin_constants';

import {setConnected} from 'src/reducers/connectedState';

const GetConfig = (): JSX.Element => {
    const {makeApiRequest, getApiState} = usePluginApi();
    const dispatch = useDispatch();

    const getConnectedUserState = () => {
        const {isLoading, data} = getApiState(Constants.pluginApiServiceConfigs.getConnectedUser.apiServiceName);
        return {isLoading, data: data as ConnectedState};
    };

    useEffect(() => {
        makeApiRequest(Constants.pluginApiServiceConfigs.getConfig.apiServiceName);
        makeApiRequest(Constants.pluginApiServiceConfigs.getConnectedUser.apiServiceName);
    }, []);

    useEffect(() => {
        const {data, isLoading} = getConnectedUserState();
        if (!isLoading && data) {
            dispatch(setConnected(data.connected));
        }
    }, [getConnectedUserState().isLoading, getConnectedUserState().data]);

    // This container is used just for making the API calls, it doesn't render anything.
    return <></>;
};

export default GetConfig;
