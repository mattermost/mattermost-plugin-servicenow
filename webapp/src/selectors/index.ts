import {ModalIds} from 'src/plugin_constants';

export const getGlobalModalState = (state: PluginState): GlobalModalState => state.globalModalReducer;

export const isAddSubscriptionModalOpen = (state: PluginState): boolean => state.globalModalReducer.modalId === ModalIds.ADD_SUBSCRIPTION;

export const isEditSubscriptionModalOpen = (state: PluginState): boolean => state.globalModalReducer.modalId === ModalIds.EDIT_SUBSCRIPTION;

export const isShareRecordModalOpen = (state: PluginState): boolean => state.globalModalReducer.modalId === ModalIds.SHARE_RECORD;

export const isCommentModalOpen = (state: PluginState): boolean => state.globalModalReducer.modalId === ModalIds.ADD_OR_VIEW_COMMENTS;

export const isUpdateStateModalOpen = (state: PluginState): boolean => state.globalModalReducer.modalId === ModalIds.UPDATE_STATE;

export const isCreateIncidentModalOpen = (state: PluginState): boolean => state.globalModalReducer.modalId === ModalIds.CREATE_INCIDENT;

export const isCreateRequestModalOpen = (state: PluginState): boolean => state.globalModalReducer.modalId === ModalIds.CREATE_REQUEST;

export const getApiRequestCompletionState = (state: PluginState): ApiRequestCompletionState => state.apiRequestCompletionReducer;
