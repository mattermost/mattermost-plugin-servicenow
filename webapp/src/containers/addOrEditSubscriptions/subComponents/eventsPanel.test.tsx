import React from 'react';

import {shallow, ShallowWrapper} from 'enzyme';

import {ModalFooter, ModalSubtitleAndError} from '@brightscout/mattermost-ui-library';

import {RecordType, RecordTypeLabelMap, SubscriptionEvents, SubscriptionType} from 'src/plugin_constants';

import EventsPanel from './eventsPanel';

const mockOnContinue = jest.fn();
const mockOnBack = jest.fn();
const mockSetSubscriptionEvents = jest.fn();

const mockSubscriptionEvents: SubscriptionEvents[] = [];

const mockChannel: DropdownOptionType = {
    label: 'mockChannelLabel',
    value: 'mockChannelValue',
};

const eventsPanelProps = {
    className: 'mockClassName',
    onBack: mockOnBack,
    onContinue: mockOnContinue,
    continueBtnDisabled: true,
    backBtnDisabled: true,
    recordType: RecordType.INCIDENT,
    subscriptionEvents: mockSubscriptionEvents,
    setSubscriptionEvents: mockSetSubscriptionEvents,
    channel: mockChannel,
    filters: [],
};

describe('Events Panel', () => {
    let component: ShallowWrapper<any, Readonly<{}>, React.Component<{}, {}, any>>;

    beforeEach(() => {
        component = shallow(
            <EventsPanel
                {...eventsPanelProps}
                subscriptionType={SubscriptionType.RECORD}
                record='mockRecord'
                error='mockError'
            />);
    });

    it('Should render correctly', () => {
        expect(component).toMatchSnapshot();
    });

    it('Should apply the passed className prop', () => {
        expect(component.hasClass(eventsPanelProps.className)).toBeTruthy();
    });

    it('Should render the events panel body correctly', () => {
        expect(component.find('Checkbox')).toHaveLength(5);
        expect(component.find('ModalSubtitleAndError')).toHaveLength(1);
        expect(component.find('ModalFooter')).toHaveLength(1);
    });

    it('Should render the events panel body correctly when subscription type is BULK', () => {
        component = shallow(
            <EventsPanel
                {...eventsPanelProps}
                subscriptionType={SubscriptionType.BULK}
                record='mockRecord'
            />);
        expect(component.find('Checkbox')).toHaveLength(6);
        expect(component.find('ModalSubtitleAndError')).toHaveLength(1);
        expect(component.find('ModalFooter')).toHaveLength(1);
    });

    it('Should render the events panel text correctly', () => {
        expect(component.text().includes('Channel')).toBeTruthy();
        expect(component.text().includes(`${eventsPanelProps.channel.label}`)).toBeTruthy();
        expect(component.text().includes('Record')).toBeTruthy();
        expect(component.text().includes('mockRecord')).toBeTruthy();
        expect(component.text().includes('Available events:')).toBeTruthy();
    });

    it('Should render the events panel text correctly on not having record and subscription type BULK', () => {
        component = shallow(
            <EventsPanel
                {...eventsPanelProps}
                subscriptionType={SubscriptionType.BULK}
                record=''
            />);
        expect(component.text().includes('Channel')).toBeTruthy();
        expect(component.text().includes(`${eventsPanelProps.channel.label}`)).toBeTruthy();
        expect(component.text().includes('Record type')).toBeTruthy();
        expect(component.text().includes(RecordTypeLabelMap[eventsPanelProps.recordType])).toBeTruthy();
        expect(component.text().includes('Available events:')).toBeTruthy();
    });

    it('Should render the error correctly', () => {
        expect(component.contains(
            <ModalSubtitleAndError error='mockError'/>,
        )).toBeTruthy();
    });

    it('Should not render the error, if error is not passed', () => {
        expect(component.contains(
            <ModalSubtitleAndError/>,
        )).toBeFalsy();
    });

    it('Should render the footer correctly', () => {
        expect(component.contains(
            <ModalFooter
                onHide={eventsPanelProps.onBack}
                onConfirm={eventsPanelProps.onContinue}
                confirmBtnText='Continue'
                cancelBtnText='Back'
                confirmDisabled={eventsPanelProps.continueBtnDisabled}
                cancelDisabled={eventsPanelProps.backBtnDisabled}
            />,
        )).toBeTruthy();
    });

    it('Should fire change event when clicked', () => {
        const clickCheckbox = (clickNumber: number, checked: boolean) => {
            // eslint-disable-next-line max-nested-callbacks
            component.find('Checkbox').forEach((node) => node.simulate('change', {target: {checked}}));
            expect(eventsPanelProps.setSubscriptionEvents).toHaveBeenCalledTimes(clickNumber);
        };

        // Click the checkbox
        clickCheckbox(5, true);
        clickCheckbox(10, false);
    });
});
