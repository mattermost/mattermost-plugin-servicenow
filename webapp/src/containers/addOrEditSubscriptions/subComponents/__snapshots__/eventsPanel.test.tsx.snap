// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Events Panel Should render correctly 1`] = `
<div
  className="modal__body wizard__secondary-panel events-panel mockClassName"
>
  <div
    className="padding-h-12 padding-v-20 wizard__body-container"
  >
    <div
      className="events-panel__prev-data border-radius-8 padding-v-10 padding-h-25 margin-bottom-25"
    >
      <div
        className="d-flex"
      >
        <div>
          <h4
            className="events-panel__prev-data-header font-14 wt-400 margin-top-15 record-header"
          >
            Channel
          </h4>
          <p
            className="events-panel__prev-data-text font-14 wt-400 margin-v-5"
          >
            mockChannelLabel
          </p>
          <h4
            className="events-panel__prev-data-header font-14 wt-400 margin-top-15 record-header"
          >
            Record
          </h4>
          <p
            className="events-panel__prev-data-text font-14 wt-400 margin-v-5"
          >
            mockRecord
          </p>
        </div>
        <div
          className="margin-left-150"
        />
      </div>
    </div>
    <label
      className="events-panel__label font-16 margin-bottom-12 wt-400"
    >
      Available events:
    </label>
    <div
      className="d-flex"
    >
      <div>
        <Checkbox
          checked={false}
          className="margin-bottom-20"
          label="State changed"
          onChange={[Function]}
        />
        <Checkbox
          checked={false}
          className="margin-bottom-20"
          label="Priority changed"
          onChange={[Function]}
        />
        <Checkbox
          checked={false}
          className="margin-bottom-20"
          label="New comment"
          onChange={[Function]}
        />
      </div>
      <div
        className="margin-left-150"
      >
        <Checkbox
          checked={false}
          className="margin-bottom-20"
          label="Assigned to changed"
          onChange={[Function]}
        />
        <Checkbox
          checked={false}
          className="margin-bottom-20"
          label="Assignment group changed"
          onChange={[Function]}
        />
      </div>
    </div>
    <ModalSubtitleAndError
      error="mockError"
    />
  </div>
  <ModalFooter
    cancelBtnText="Back"
    cancelDisabled={true}
    confirmBtnText="Continue"
    confirmDisabled={true}
    onConfirm={[MockFunction]}
    onHide={[MockFunction]}
  />
</div>
`;
